Move tabs to sidebar

FRCassarino opened the pull request:
Moves group tabs to the sidebar. 

<img width="1352" alt="image" src="https://user-images.githubusercontent.com/6242616/190207544-05bb1eb2-7125-4933-9d22-debd17c1d7bb.png">
<img width="1352" alt="image" src="https://user-images.githubusercontent.com/6242616/190207586-086d3e04-dbe0-4fe2-8bbf-9b03c944ad9f.png">
<img width="1270" alt="image" src="https://user-images.githubusercontent.com/6242616/190207609-9a7c2191-6550-4f65-8866-fe74c175feaa.png">
<img width="595" alt="image" src="https://user-images.githubusercontent.com/6242616/190207715-d1542864-f61a-4d4e-a224-7b9453c09108.png">
<img width="560" alt="image" src="https://user-images.githubusercontent.com/6242616/190207767-ed49d767-8fbe-4321-8a70-b3f2b1f74ec2.png">
<img width="515" alt="image" src="https://user-images.githubusercontent.com/6242616/190207779-f84b3183-f90e-401b-b62c-1af8f1283c42.png">

FRCassarino commented:
Still has a couple of visual bugs that I'll need to fix before shipping, but it should be shippable after that if we agree on it.

@akrolsmir @jahooma Is there anything else you'd like added before we ship this?

FRCassarino commented on web/pages/group/[...slugs]/index.tsx:
```diff
@@ -196,85 +198,118 @@ export default function GroupPage(props: {
-              <Linkify text={group.about} />
-            </div>
+    <>
+      <header className="sticky top-0 z-50 w-full pb-2 md:hidden lg:col-span-12">
```
https://user-images.githubusercontent.com/6242616/190210177-4cb11d60-8792-4af9-be01-35af7c234f69.mov

Does anybody have a good idea for how to fix this? Tailwinds 'sticky' position uses the browser window as its position reference, so ContractsSearchFilters overlaps the header. What I want is the same behaviour, but with the component stopping at the header instead.

jahooma replied:
Hmm, good question. I'm not sure. There might be a way to add an offset to the search bar's sticky y-offset.

Alternatively, you could just pass an option to disable the sticky search bar for groups on mobile.

jahooma replied:
One small suggestion might be to extract this header to a new component in this file.

gappleto97 commented:
Here's a market for this: https://manifold.markets/MarketManagerBot/how-many-days-after-september-1st-w-7a49d7fb4b06

jahooma commented on web/components/nav/group-nav-bar.tsx:
```diff
@@ -0,0 +1,94 @@
+  )
+
+  return (
+    <button onClick={() => props.onClick(item.key ?? '#')}>
+      <a
+        className={clsx(
+          'block w-full py-1 px-3 text-center hover:bg-indigo-200 hover:text-indigo-700',
+          currentPage === item.key && 'bg-gray-200 text-indigo-700'
+        )}
+        onClick={track}
+      >
+        {item.icon && <item.icon className="my-1 mx-auto h-6 w-6" />}
+        {item.name}
+      </a>
+    </button>
```
I'm not sure if this kind of link works in all cases. For example, if someone middle-clicks the button, would it open the link in a new tab? Or, if you tab to focus-select it and hit enter, does it navigate (and also trigger the track function)?

You might have to use the nextjs `<Link>` wrapper and set `href`, or use `SiteLink`.

jahooma commented on web/components/nav/group-sidebar.tsx:
```diff
@@ -0,0 +1,87 @@
+    }
+  )
+
+export default function GroupSidebar(props: {
```
We don't typically use default exports, except for pages. It's good to be consistent so you always know how to import anything.

Two reasons for this convention: 1. can easily export more things from a file without updating all references 2. Renaming an export will automatically rename it everywhere it is referenced, which is not true for default exports.

jahooma commented on web/pages/group/[...slugs]/index.tsx:
```diff
@@ -196,85 +198,118 @@ export default function GroupPage(props: {
+            className="sticky top-0 hidden divide-gray-300 self-start pl-2 lg:col-span-2 lg:flex"
+            onClick={onSidebarClick}
+            joinOrAddQuestionsButton={joinOrAddQuestionsButton}
+            currentKey={sidebarPages[sidebarIndex].key}
```
This should probably be represented in the URL, I think. Also could allow you to go back and forth.

A long time ago, you could go to `/group/JamesGroup/leaderboard` and it would show the leaderboard, etc.

Actually, now that these are not tabs, I think it's even more expected that the distinct pages of markets, about, leaderboard are different URLs. You can either change this PR, or follow up with the change.

jahooma replied:
You might need to use nextjs catchall routes: https://nextjs.org/docs/routing/dynamic-routes#optional-catch-all-routes

jahooma replied:
Actually, it might be better to make each a separate page as a new file under `pages/group/`.

E.g. `pages/group/my-group/markets`. Would be much better organization

jahooma reviewed (approved):
Sweet! This is really cool, Fede! Glad to see this groups overhaul come together quickly.

I tend to favor the approach of shipping it and seeing how it goes, with follow ups as needed. With that in mind, I approve this PR.

**Thoughts**
- It seems to me that on desktop the name of the group is not that prominent, so that someone wouldn't necessarily realize this is a sub-Manifold site.
One idea I had was to move the group name below the logo, which makes it more prominent. See image below.
- I don't like different indentation levels for sidebar items.
- Would be great if clicking notifications kept you within the group. Might be encoded as `/notifications?g=group-slug`. Not sure though.
- I'm not sure about the mobile navigation, seems like something to iterate on.
- I do like the sticky group name and back arrow on mobile. Great solution.

![mockup groups](https://user-images.githubusercontent.com/1593821/190322765-02f84ae8-4f26-440c-8603-6dde911cdec4.png)

FRCassarino commented:
I addressed all the feedback items, except for the following ones which are all related:

> This should probably be represented in the URL, I think. Also could allow you to go back and forth.
> 
> A long time ago, you could go to `/group/JamesGroup/leaderboard` and it would show the leaderboard, etc.
> 
> Actually, now that these are not tabs, I think it's even more expected that the distinct pages of markets, about, leaderboard are different URLs. You can either change this PR, or follow up with the change.

> Actually, it might be better to make each a separate page as a new file under `pages/group/`.
> 
> E.g. `pages/group/my-group/markets`. Would be much better organization


> I'm not sure if this kind of link works in all cases. For example, if someone middle-clicks the button, would it open the link in a new tab? Or, if you tab to focus-select it and hit enter, does it navigate (and also trigger the track function)?
> 
> You might have to use the nextjs `<Link>` wrapper and set `href`, or use `SiteLink`.

The same changes should address all of those, but I'll do it in a separate PR. In the meantime I'm merging this version.
